src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -0,0 +1,283 @@
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (protectedHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(protectedHeaders));
+            }
+
+            if (unprotectedHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(unprotectedHeaders));
+            }
+
+            KeyType keyType = key switch
+            {
+                ECDsa => KeyType.ECDsa,
+                RSA => KeyType.RSA,
+                _ => throw new CryptographicException(SR.Format(SR.Sign1SignUnsupportedKey, key.GetType()))
+            };
+
+            ThrowIfDuplicateLabels(protectedHeaders, unprotectedHeaders);
+
+            ReadOnlyMemory<byte> encodedAlg = GetCoseAlgorithmFromBuckets(protectedHeaders, unprotectedHeaders);
+
+            int? algorithmHeader = DecodeCoseAlgorithmHeader(encodedAlg);
+            Debug.Assert(algorithmHeader.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
+
+            HashAlgorithmName hashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(algorithmHeader.Value, keyType);
+
+            byte[] encodedProtetedHeaders = protectedHeaders.Encode(mustReturnEmptyBstrIfEmpty: true);
+            byte[] toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtetedHeaders, content);
+
+            byte[] signature = keyType == KeyType.ECDsa ?
+                SignWithECDsa((ECDsa)key, toBeSigned, hashAlgorithm) :
+                SignWithRSA((RSA)key, toBeSigned, hashAlgorithm);
+
+            return SignCore(encodedProtetedHeaders, unprotectedHeaders.Encode(), signature, content, isDetached);
+        }
+
+        // Validate duplicate labels https://datatracker.ietf.org/doc/html/rfc8152#section-3.
+        internal static void ThrowIfDuplicateLabels(CoseHeaderMap @protected, CoseHeaderMap unprotected)
+        {
+            foreach ((CoseHeaderLabel Label, ReadOnlyMemory<byte>) header in @protected)
+            {
+                if (unprotected.TryGetEncodedValue(header.Label, out _))
+                {
+                    throw new CryptographicException(SR.Sign1SignHeaderDuplicateLabels);
+                }
+            }
+        }
+
+        private static byte[] SignCore(ReadOnlySpan<byte> encodedProtectedHeader, ReadOnlySpan<byte> encodedUnprotectedHeader, ReadOnlySpan<byte> signature, ReadOnlySpan<byte> content, bool isDetached)
+        {
+            var writer = new CborWriter();
+            writer.WriteTag(Sign1Tag);
+            writer.WriteStartArray(Sign1ArrayLegth);
+            writer.WriteByteString(encodedProtectedHeader);
+            writer.WriteEncodedValue(encodedUnprotectedHeader);
+            if (isDetached)
+            {
+                writer.WriteNull();
+            }
+            else
+            {
+                writer.WriteByteString(content);
+            }
+            writer.WriteByteString(signature);
```

Comment:
Personally, I think a bit more whitespace will help readibility here. 

There're no hard rules here, but in "Jeremy always shares his opinion... whether people want it or not." style:  I, personally, always have a blank line (or line of just a `{`) before an if (except in argument validation when using braceless ifs).  And a blank line (or `}`) as the line after a closing brace.  All of the code before the if goes together, like sentences in a paragraph.  Then a blank line for "we're doing something a little different now" (hit enter (one more time) for the paragraph start), the conditional code, then once we're done with the conditional code it's another blank line because we're doing something else...

```suggestion
            writer.WriteEncodedValue(encodedUnprotectedHeader);
            
            if (isDetached)
            {
                writer.WriteNull();
            }
            else
            {
                writer.WriteByteString(content);
            }
            
            writer.WriteByteString(signature);
```

Reply:
Ok, sounds reasonable.